### ISSUE TITLE
Cannot make Docker Compose based App Services with new App Service resources

### ISSUE THREAD

**Issue by alexgagnon** [terraform-providers/terraform-provider-azurerm#16290]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

1.1.7

### AzureRM Provider Version

3.0.2

### Affected Resource(s)/Data Source(s)

azurerm_[linux|windows]_web_app

### Terraform Configuration Files

```hcl
resource "azurerm_linux_web_app" "as" {
  name = "example"
  resource_group_name = "rg-example"
  location = "westus2"
  service_plan_id = azurerm_service_plan.example.id

  site_config {
    application_stack {
      docker_image = "COMPOSE|${filebase64("example.docker-compose.yml")}"
      docker_image_tag = "?"
    }   
  }
}
```

### Debug Output/Panic Output

```shell
n/a
```

### Expected Behaviour

Should be able to create a multi-container linux web application based on the provided Docker Compose file

### Actual Behaviour

Tries to convert to an invalid linux_fx_version, which spits out the following:

```
│ Error: creating Linux Web App: (Site Name "as" / Resource Group "rg-example"): web.AppsClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="BadRequest" Message="The parameter LinuxFxVersion has an invalid value." Details=[{"Message":"The parameter LinuxFxVersion has an invalid value."},{"Code":"BadRequest"},{"ErrorEntity":{"Code":"BadRequest","ExtendedCode":"01007","Message":"The parameter LinuxFxVersion has an invalid value.","MessageTemplate":"The parameter {0} has an invalid value.","Parameters":["LinuxFxVersion"]}}]
│ 
│   with azurerm_linux_web_app.as
│   on sonarqube.tf line 19, in resource "azurerm_linux_web_app" "as":
│   19: resource "azurerm_linux_web_app" "as" {
│ 
│ creating Linux Web App: (Site Name "as" / Resource Group
│ "rg-dexample"): web.AppsClient#CreateOrUpdate: Failure sending
│ request: StatusCode=400 -- Original Error: Code="BadRequest" Message="The
│ parameter LinuxFxVersion has an invalid value." Details=[{"Message":"The
│ parameter LinuxFxVersion has an invalid
│ value."},{"Code":"BadRequest"},{"ErrorEntity":{"Code":"BadRequest","ExtendedCode":"01007","Message":"The
│ parameter LinuxFxVersion has an invalid value.","MessageTemplate":"The
│ parameter {0} has an invalid value.","Parameters":["LinuxFxVersion"]}}]
```

### Important Factoids

n/a

### References

I'm not sure if migrating from azurerm provider from v2 to v3 removed the ability to create multi-container applications, but the previous example of creating these resources don't seem to work anymore

[https://github.com/hashicorp/terraform-provider-azurerm/blob/main/examples/app-service/docker-compose/main.tf](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/examples/app-service/docker-compose/main.tf
)

In v3 when using [linux|window]_web_app, we provide specific docker_image and docker_image_tag values, which are converted into the linux_fx_version field. However there is no COMPOSE option to match against: see [here](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/appservice/helpers/web_app_schema.go#L2962-L2964) and [here](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/appservice/helpers/fx_strings.go#L45-L50).

EDIT: line numbers in second link have changed, here's the code in question:

```go
			if winAppStack.DockerContainerName != "" {
				if winAppStack.DockerContainerRegistry != "" {
					expanded.WindowsFxVersion = utils.String(fmt.Sprintf("DOCKER|%s/%s:%s", winAppStack.DockerContainerRegistry, winAppStack.DockerContainerName, winAppStack.DockerContainerTag))
				} else {
					expanded.WindowsFxVersion = utils.String(fmt.Sprintf("DOCKER|%s:%s", winAppStack.DockerContainerName, winAppStack.DockerContainerTag))
				}
			}
```

and

```go
			if linuxAppStack.DockerImage != "" {
				expanded.LinuxFxVersion = utils.String(fmt.Sprintf("DOCKER|%s:%s", linuxAppStack.DockerImage, linuxAppStack.DockerImageTag))
			}
```

The tutorial provided by Azure here doesn't even list referencing a Container Registry... https://docs.microsoft.com/en-us/azure/app-service/quickstart-multi-container

**Comment by KoenR3** [comment]
This is a major blocking issue for us as it removes core Azure functionality

**Comment by nicole-ashley** [comment]
I would update all my app service resources right away if this was fixed 😕 Until then we'll be stuck on version 3 with countless warning messages.

**Comment by Raraku** [comment]
Hi, 
Is there any fix relating to this coming soon?

**Comment by clowa** [comment]
Hi there,
am I missing something or is the `docker-compose` deployment not supported by the new AzureRM provider? I read throw the [terraform registry](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/linux_web_app) of the `azurerm_linux_web_app` resource but didn't find any configuration option to use `docker-compose` - did I miss something or do we have to use the deprecated `azurerm_app_service` resource? 

Really looking forward to get an update on this !

**Comment by xiaxyi** [comment]
Thanks @alexgagnon, I'll check to see how to add the compose option.

**Comment by xiaxyi** [comment]
Hi @alexgagnon and guys, I added the docker compose option in the pr. you can track the status there
https://github.com/hashicorp/terraform-provider-azurerm/pull/18053

**Comment by alicyn** [comment]
@xiaxyi is there any update on this? I am facing the same issue here.

**Comment by xiaxyi** [comment]
Thanks @alicyn , The fix is included in the mentioned pr, we are doing the internal review, will let you know once the code is ready to be released. Thanks for your patience!

**Comment by valourus** [comment]
I need  this for my school project

**Comment by bdassow-nc** [comment]
Anything new on this?
It´s somehow quite challenging foreced to use a workaround because of a missing properties, thats existed in the Version before.

**Comment by jbowes68** [comment]
> Anything new on this? It´s somehow quite challenging foreced to use a workaround because of a missing properties, thats existed in the Version before.

what workaround are you using for this?

**Comment by bdassow-nc** [comment]
@jbowes68 
I´m using the Azure/azapi provider to update the deployed resource.

```hcl
resource "azapi_update_resource" "update_linux_web_app" {
  resource_id = var.linux_web_app_id
  type        = "Microsoft.Web/sites@2022-03-01"
  body = jsonencode({
    properties = {
      "siteConfig" = {
        "linuxFxVersion" = "COMPOSE|${base64encode(file("docker-compose.yml"))}"
      }
    }
  })
}
```

**Comment by swissbuechi** [comment]
> @jbowes68 I´m using the Azure/azapi provider to update the deployed resource.
> 
> ```terraform
> resource "azapi_update_resource" "update_linux_web_app" {
>   resource_id = var.linux_web_app_id
>   type        = "Microsoft.Web/sites@2022-03-01"
>   body = jsonencode({
>     properties = {
>       "siteConfig" = {
>         "linuxFxVersion" = "COMPOSE|${base64encode(file("docker-compose.yml"))}"
>       }
>     }
>   })
> }
> ```
@bdassow-nc 
Thanks for your workaround.
How do you deploy a new app service?

**Comment by bdassow-nc** [comment]
@swissbuechi 
I´m using the regular terraform resource for my destination app service (linux/windows) and through the azapi resource right after with an depends_on = []

But in my case I switched from App Service to Container Instance, as the docker-compose file was way to big for app service . There is an limit of 4k chars.

**Comment by samrocksc** [comment]
Another person looking for a timeline on this fix!  Am interested.

**Comment by xiaxyi** [comment]
Thanks all for the active discussion here, the support for docker compose is included in the pr https://github.com/hashicorp/terraform-provider-azurerm/pull/20735

**Comment by xiaxyi** [comment]
Thanks Guys for the active discussion here, the support of docker compose file is included in the mentioned PR.

**Comment by samrocksc** [comment]
do we by chance have a working example of this?

**Comment by clowa** [comment]
> > @jbowes68 I´m using the Azure/azapi provider to update the deployed resource.
> > ```terraform
> > resource "azapi_update_resource" "update_linux_web_app" {
> >   resource_id = var.linux_web_app_id
> >   type        = "Microsoft.Web/sites@2022-03-01"
> >   body = jsonencode({
> >     properties = {
> >       "siteConfig" = {
> >         "linuxFxVersion" = "COMPOSE|${base64encode(file("docker-compose.yml"))}"
> >       }
> >     }
> >   })
> > }
> > ```
> 
> @bdassow-nc Thanks for your workaround. How do you deploy a new app service?

@samrocksc, you could use the above example from @swissbuechi using the [`azapi provider`](https://registry.terraform.io/providers/Azure/azapi/latest)

The following example works:

```hcl
#######################################
# Provider

terraform {
  required_version = "~>1.5"

  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
      version = "~> 3.63"
    }
    azapi = {
      source  = "Azure/azapi"
      version = "~> 1.7"
    }
  }
}

provider "azurerm" {
  features {}
}

provider "azapi" {}

#######################################
# Variables

variable "global_prefix" {
  description = "Unique identifer for ressurce naming"
  type        = string
}

variable "location" {
  description = "azure region name"
  type        = string
}

variable "tags" {
  description = "Tags to apply on any resource within this project"
  type        = map(string)
  default = {
    deploymentMethod = "terraform"
  }
}

#######################################
# Resources

resource "azurerm_resource_group" "this" {
  name     = "${var.global_prefix}-rg"
  location = var.location
  tags     = var.tags
}

resource "azurerm_service_plan" "this" {
  name                = "${var.global_prefix}-asp"
  resource_group_name = azurerm_resource_group.this.name
  location            = azurerm_resource_group.this.location
  tags                = var.tags

  os_type  = "Linux"
  sku_name = "B2"
}

resource "azurerm_linux_web_app" "this" {
  name                = "${var.global_prefix}-app"
  resource_group_name = azurerm_resource_group.this.name
  location            = azurerm_resource_group.this.location

  #https_only              = true

  service_plan_id = azurerm_service_plan.this.id

  app_settings = {
    DOCKER_ENABLE_CI                = "true"
    DOCKER_REGISTRY_SERVER_URL      = "https://index.docker.io/v1"
    WEBSITE_ENABLE_SYNC_UPDATE_SITE = "true"
  }

  site_config {}
}

# App Service Resource currently doesn't support to set docker compose
# See: https://github.com/hashicorp/terraform-provider-azurerm/issues/16290
resource "azapi_update_resource" "update_linux_web_app" {
  resource_id = azurerm_linux_web_app.this.id
  type        = "Microsoft.Web/sites@2022-03-01"
  body = jsonencode({
    properties = {
      "siteConfig" = {
        "linuxFxVersion" = "COMPOSE|${base64encode(file("./docker-compose.yaml"))}"
      }
    }
  })
}
```

... my `docker-compose.yaml` looks like this:

```yaml
version: "3.3"

services:
  whoami:
    image: "traefik/whoami"
    container_name: "simple-service"
    ports:
      - "80:80"

```

Currently Azure App Service has certain restrictions on custom containers via `docker compose`, you can more about them [here](https://learn.microsoft.com/en-us/azure/app-service/configure-custom-container).

**Comment by rcskosir** [comment]
PR https://github.com/hashicorp/terraform-provider-azurerm/pull/20735 has been superseded by https://github.com/hashicorp/terraform-provider-azurerm/pull/22003, which has merged. Please upgrade to the latest provider (3.63) if you are able, to let us know if this has resolved your issue.
Thank you

**Comment by hoangtoni** [comment]
> @rcskosir I finally got it working! Thanks a lot for the follow up! I think the only thing left at this point is how to set the environment variables _inside_ of the docker-compose file, or how to update the image that it's trying to get

@samrocksc could you give provide informaiton, on how you got it working? Is it now closed? I tried the "expected" way and it did not work.

**Comment by Fabe1997** [comment]
Still broken in 2024.

**Comment by gionkunz** [comment]
I also needed this and additionally I needed to configure an container registry:

```
resource "azurerm_linux_web_app" "wa-website-1" {
  app_settings = {
    DOCKER_ENABLE_CI                    = true
    WEBSITES_ENABLE_APP_SERVICE_STORAGE = true
    DOCKER_REGISTRY_SERVER_URL          = "https://srxxxxx.azurecr.io"
  }
  https_only          = true
  location            = azurerm_resource_group.rg-website-1.location
  name                = "wa-${var.name}-website-${var.environment}"
  resource_group_name = azurerm_resource_group.rg-website-1.name
  service_plan_id     = data.azurerm_service_plan.sp-global.id
  backup {
    name                = "xxxxx"
    storage_account_url = "https://${azurerm_storage_account.st-website-1.name}.blob.core.windows.net/${azurerm_storage_container.sc-website-backup-1.name}?${data.azurerm_storage_account_sas.sas-website-1.sas}"
    schedule {
      frequency_interval       = 1
      frequency_unit           = "Day"
      keep_at_least_one_backup = true
    }
  }
  identity {
    type = "SystemAssigned"
  }
  logs {
    http_logs {
      file_system {
        retention_in_days = 14
        retention_in_mb   = 35
      }
    }
  }
  site_config {
    always_on                               = true
    container_registry_use_managed_identity = true
    ftps_state                              = "FtpsOnly"
  }
}

data "local_file" "dockercompose" {
  filename = "../../docker-compose.azure.${var.environment}.yml"
}

resource "azapi_update_resource" "update_linux_web_app" {
  resource_id = azurerm_linux_web_app.wa-website-1.id
  type        = "Microsoft.Web/sites@2022-03-01"
  body = jsonencode({
    properties = {
      siteConfig = {
        linuxFxVersion = "COMPOSE|${data.local_file.dockercompose.content_base64}"
        acrUseManagedIdentityCreds = true
      }
    }
  })
```
